File: src/main/java/me/lokka30/levelledmobs/misc/VersionInfo.java
/*
 * Copyright (c) 2020-2021  lokka30. Use of this source code is governed by the GNU AGPL v3.0 license that can be found in the LICENSE.md file.
 */

package me.lokka30.levelledmobs.misc;

import java.io.InvalidObjectException;
import java.util.LinkedList;
import java.util.List;

/**
 * A custom implementation for comparing program versions
 *
 * @author stumper66
 * @since 2.6.0
 */
public class VersionInfo implements Comparable<VersionInfo> {
    public VersionInfo(final String version) throws InvalidObjectException {
        if (version == null) throw new NullPointerException("version can't be null");

        this.versionStr = version;
        String[] split = version.split("\\.");
        this.thisVerSplit = new LinkedList<>();
        for (final String numTemp : split){
            if (!Utils.isDouble(numTemp)) throw new InvalidObjectException("Version can only contain numbers and periods");
            int intD = Integer.parseInt(numTemp);
            thisVerSplit.add(intD);
        }

        for (int i = 4; i < thisVerSplit.size(); i++)
            thisVerSplit.add(0);
    }

    final private String versionStr;
    final List<Integer> thisVerSplit;

    @Override
    public boolean equals(final Object o) {
        if (o == null) return false;
        if (o == this) return true;
        if (!(o instanceof VersionInfo)) return false;

        final VersionInfo otherVersion = (VersionInfo) o;
        return this.versionStr.equals(((VersionInfo) o).getVersion());
    }

    @Override
    public String toString() {
        return this.versionStr;
    }

    public String getVersion(){
        return this.versionStr;
    }

    @Override
    public int compareTo(final VersionInfo v) {
        for (int i = 0; i < 4; i++) {

            if (v.thisVerSplit.size() <= i && this.thisVerSplit.size() - 1 <= i)
                break;

                // if one has extra digits we'll assume that one is newer
            else if (v.thisVerSplit.size() <= i)
                return 1;
            else if (this.thisVerSplit.size() <= i)
                return -1;

            final int compareInt = v.thisVerSplit.get(i);
            final int thisInt = this.thisVerSplit.get(i);

            if (thisInt > compareInt)
                return 1;
            else if (thisInt < compareInt)
                return -1;
        }

        return 0;
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/misc/VersionInfo.kt
package io.github.arcaneplugins.levelledmobs.misc

import java.io.InvalidObjectException
import io.github.arcaneplugins.levelledmobs.util.Utils

/**
 * A custom implementation for comparing program versions
 *
 * @author stumper66
 * @since 2.6.0
 */
class VersionInfo(
    versionInput: String
) {
    private var thisVerSplit = mutableListOf<Int>()
    val version: String

    init {
        val buildNum = versionInput.indexOf("build") // 26.1.1.build.15
        version = if (buildNum > 0)
            versionInput.substring(0, buildNum)
        else
            versionInput

        val split = version.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (numTemp in split) {
            if (!Utils.isDouble(numTemp))
                throw InvalidObjectException("Version can only contain numbers and periods")

            val intD = numTemp.toInt()
            thisVerSplit.add(intD)
        }

        repeat(4 - thisVerSplit.size) {
            thisVerSplit.add(0)
        }
    }

    fun isLessThan(version: String): Boolean {
        return isLessThan(VersionInfo(version))
    }

    fun isLessThan(version: VersionInfo): Boolean {
        return compare(version) == CompareResult.LESS_THAN
    }

    fun isLessThanOrEquals(version: String): Boolean {
        return isLessThanOrEquals(VersionInfo(version))
    }

    fun isLessThanOrEquals(version: VersionInfo): Boolean {
        val result = compare(version)
        return result == CompareResult.LESS_THAN ||
                result == CompareResult.EQUAL
    }

    fun isGreaterThan(version: String): Boolean {
        return isGreaterThan(VersionInfo(version))
    }

    fun isGreaterThan(version: VersionInfo): Boolean {
        return compare(version) == CompareResult.GREATER_THAN
    }

    fun isGreaterThanOrEqual(version: String): Boolean {
        return isGreaterThanOrEqual(VersionInfo(version))
    }

    fun isGreaterThanOrEqual(version: VersionInfo): Boolean {
        val result = compare(version)
        return result == CompareResult.GREATER_THAN ||
                result == CompareResult.EQUAL
    }

    private fun compare(other: VersionInfo) : CompareResult {
        for (i in 0..3) {
            if (other.thisVerSplit.size <= i && thisVerSplit.size - 1 <= i)
                break

            else if (other.thisVerSplit.size <= i)
                return CompareResult.GREATER_THAN
            else if (thisVerSplit.size <= i)
                return CompareResult.LESS_THAN

            val compareInt: Int = other.thisVerSplit[i]
            val thisInt = thisVerSplit[i]

            if (thisInt > compareInt)
                return CompareResult.GREATER_THAN
            else if (thisInt < compareInt)
                return CompareResult.LESS_THAN
        }

        return CompareResult.EQUAL
    }

    enum class CompareResult{
        LESS_THAN,
        EQUAL,
        GREATER_THAN
    }

    override fun toString(): String {
        return version
    }
}